resources/prosody-plugins/mod_speakerstats_component.lua

```diff
@@ -0,0 +1,189 @@
+            newDominantSpeaker:setIsDominantSpeaker(true);
+        end
+
+        room.speakerStats['dominantSpeakerId'] = occupant.jid;
+    end
+
+    return true
+end
+
+--- Start SpeakerStats implementation
+local SpeakerStats = {};
+SpeakerStats.__index = SpeakerStats;
+
+function new_SpeakerStats(nick)
+    return setmetatable({
+        totalDominantSpeakerTime = 0;
+        _dominantSpeakerStart = nil;
+        _isDominantSpeaker = false;
+        nick = nick;
+        displayName = nil;
+    }, SpeakerStats);
+end
+
+-- Changes the dominantSpeaker data for current occupant
+-- saves start time if it is new dominat speaker
+-- or calculates and accumulates time of speaking
+function SpeakerStats:setIsDominantSpeaker(isNowDominantSpeaker)
+    log("debug",
+        "set isDominant %s for %s", tostring(isNowDominantSpeaker), self.nick);
+
+    if not self._isDominantSpeaker and isNowDominantSpeaker then
+        self._dominantSpeakerStart = socket.gettime()*1000;
+    elseif self._isDominantSpeaker and not isNowDominantSpeaker then
+        local now = socket.gettime()*1000;
+        local timeElapsed = math.floor(now - (self._dominantSpeakerStart or 0));
+
+        self.totalDominantSpeakerTime
+            = self.totalDominantSpeakerTime + timeElapsed;
+        self._dominantSpeakerStart = nil;
+    end
+
+    self._isDominantSpeaker = isNowDominantSpeaker;
+end
+--- End SpeakerStats
+
+-- create speakerStats for the room
+function room_created(event)
+    local room = event.room;
+    room.speakerStats = {};
+end
+
+-- Create SpeakerStats object for the joined user
+function occupant_joined(event)
+    local room = event.room;
+    local occupant = event.occupant;
+    local nick = jid_resource(occupant.nick);
+
+    if room.speakerStats then
+        -- lets send the current speaker stats to that user, so he can update
+        -- its local stats
+        if next(room.speakerStats) ~= nil then
+            local users_json = {};
+            for jid, values in pairs(room.speakerStats) do
+                -- skip reporting those without a nick('dominantSpeakerId')
+                -- and skip focus if sneaked into the table
+                if values.nick ~= nil and values.nick ~= 'focus' then
+                    local resultSpeakerStats = {};
+                    local totalDominantSpeakerTime
+                        = values.totalDominantSpeakerTime;
+
+                    -- before sending we need to calculate current dominant speaker
+                    -- state
+                    if values._dominantSpeakerStart ~= nil then
+                        local timeElapsed = math.floor(
+                            socket.gettime()*1000
+                                - (values._dominantSpeakerStart or 0));
+                        totalDominantSpeakerTime = totalDominantSpeakerTime
+                            + timeElapsed;
+                    end
+
+                    resultSpeakerStats.displayName = values.displayName;
```

Comment:
I'm wondering why do we need the display name? I guess any client will be maintaining jid -> display name mapping any way.

Reply:
Cause, the results can be used to be pushed to another service. For example publishing speaker stats in slack room.

Reply:
I see! It's OK then!